src/utils/marketsAndNetworksConfig.ts

```diff
@@ -143,6 +145,110 @@ export const isFeatureEnabled = {
+}
+
+class StaticJsonRpcBatchProvider extends ethersProviders.JsonRpcBatchProvider {
+  async detectNetwork(): Promise<Network> {
+    let network = this.network;
+    if (network == null) {
+      network = await super.detectNetwork();
+
+      if (!network) {
+        logger.throwError('no network detected', Logger.errors.UNKNOWN_ERROR, {});
+      }
+
+      // If still not set, set it
+      if (this._network == null) {
+        // A static network does not support "any"
+        defineReadOnly(this, '_network', network);
+
+        this.emit('network', network, null);
+      }
+    }
+    return network;
+  }
+}
+
+/**
+ * The provider will rotate rpcs on error.
+ */
+class RotationProvider extends ethersProviders.BaseProvider {
+  readonly providers: StaticJsonRpcBatchProvider[];
+  private currentProviderIndex = 0;
+  private lastRotation = 0;
+  private retriedFirst = false;
+
+  constructor(urls: string[], chainId: number) {
+    super(chainId);
+    this.providers = urls.map((url) => new StaticJsonRpcBatchProvider(url, chainId));
+  }
+
+  /**
+   * Add a delay after completing a rotation, so we don't ddos a provider
+   */
+  async delayRotation() {
+    const now = new Date().getTime();
+    const diff = now - this.lastRotation;
+    if (diff < 5000) sleep(5000 - diff);
+  }
+
+  // To prioritize private endpoints and reduce the splitting of queries, the first RPC is tried twice
+  // If all RPCs fail, then delay and loop again
+  private async rotateUrl() {
+    if (this.currentProviderIndex === 0) {
+      if (this.retriedFirst) {
+        this.currentProviderIndex += 1;
+      } else {
+        this.retriedFirst = true;
+      }
+    } else if (this.currentProviderIndex === this.providers.length - 1) {
+      await this.delayRotation();
+      this.currentProviderIndex = 0;
+      this.retriedFirst = false;
+      this.lastRotation = new Date().getTime();
+    } else {
+      this.currentProviderIndex += 1;
+    }
+  }
+
+  async detectNetwork(): Promise<Network> {
+    const index = this.currentProviderIndex;
+    try {
+      return await this.providers[index].detectNetwork();
+    } catch (e) {
+      console.log(e.message);
+      await this.rotateUrl();
+      return this.detectNetwork();
+    }
+  }
+
+  // eslint-disable-next-line
+  async send(method: string, params: Array<any>): Promise<any> {
+    const index = this.currentProviderIndex;
+    try {
+      return await this.providers[index].send(method, params);
+    } catch (e) {
+      console.log(e.message);
+      await this.rotateUrl();
+      return this.send(method, params);
+    }
+  }
+
+  // eslint-disable-next-line
+  async perform(method: string, params: any): Promise<any> {
+    const index = this.currentProviderIndex;
+    try {
+      return await this.providers[index].perform(method, params);
+    } catch (e) {
+      console.log(e.message);
+      await this.rotateUrl();
+      return await this.perform(method, params);
+    }
+  }
+}
```

Comment:
In general, I think all these class methods warrant a JSDoc comment - will help explain this round robin method a bit for anyone new.

Reply:
Added comments to explain how the provider works in detail